cli/src/commands/run.rs

```diff
@@ -14,12 +14,262 @@
+        // Watchman requires a multithreaded runtime, so just reuse it.
+        return runtime::Handle::current();
+    }
+    if jobs == 1 {
+        builder.max_blocking_threads(1);
+    } else {
+        builder.max_blocking_threads(jobs);
+    }
+    let rt = builder.build().unwrap();
+    rt.handle().clone()
+}
+
+/// A commit stored under `.jj/run/`
+// TODO: Create a caching backend, which creates these on a dedicated thread or
+// threadpool.
+struct StoredCommit {
+    /// Obfuscated name for an easier lookup. If a tree/directory its not set
+    name: Option<String>,
+    /// The respective commit unmodified.
+    commit: Commit,
+    /// The output directory of the commit, contains stdout and stderr for it
+    output_dir: PathBuf,
+    /// Self-explanatory
+    working_copy_dir: PathBuf,
+    state_dir: PathBuf,
+    /// The `stdout` of the commit
+    stdout: File,
+    /// The `stderr` of the commit
+    stderr: File,
+}
+
+impl StoredCommit {
+    fn new(
+        name: Option<String>,
+        commit: &Commit,
+        output_dir: PathBuf,
+        working_copy_dir: PathBuf,
+        state_dir: PathBuf,
+        stdout: File,
+        stderr: File,
+    ) -> Self {
+        Self {
+            name,
+            commit: commit.clone(),
+            output_dir,
+            working_copy_dir,
+            state_dir,
+            stdout,
+            stderr,
+        }
+    }
+}
+
+fn create_output_files(id: CommitId, path: &PathBuf) -> Result<(File, File), io::Error> {
+    let stdout_path = path.join("output").join(format!("stdout.{id}"));
+    let stderr_path = path.join("output").join(format!("stderr.{id}"));
+    let mut file_options = OpenOptions::new();
+    let stdout = file_options.write(true).open(stdout_path)?;
+    let stderr = file_options.write(true).open(stderr_path)?;
+    Ok((stdout, stderr))
+}
+
+fn create_working_copies(
+    repo_path: &Path,
+    commits: &[Commit],
+) -> Result<Vec<StoredCommit>, io::Error> {
+    let mut results = vec![];
+    // TODO: should be stored in a backend and not hardcoded.
+    let base_path = repo_path.join("run").join("default");
+    for commit in commits {
+        let name = to_wc_name(commit.tree_id());
+
+        let (output_dir, working_copy_dir, state_dir) = create_working_copy_paths(&base_path)?;
+        let (stdout, stderr) = create_output_files(commit.id(), &base_path)?;
+
+        let stored_commit = StoredCommit::new(
+            Some(name),
+            commit,
+            output_dir,
+            working_copy_dir,
+            state_dir,
+            stdout,
+            stderr,
+        );
+        results.push(stored_commit);
+    }
+    Ok(results)
+}
+
+/// The result of a single command invocation in `run_inner`.
+struct RunJobResult {
+    /// The old `CommitId` of the commit.
+    old_id: CommitId,
+    /// The new `CommitId` for the commit.
+    rewritten_id: CommitId,
+    /// The new tree generated from the commit.
+    new_tree: Tree,
+}
```

Comment:
Does this make sense?

Reply:
question: Under what circumstance is `new_tree` different from the tree for the `rewritten_id` commit?

Reply:
It never will be different, but to make my question a bit preciser should the caller or the callee do the actual "rewrite" of the commits tree?

Reply:
I think it makes sense for the callee/result of the individual job to construct the new tree for an individual commit (considering that working-copy snapshotting happens as part of most commands anyways, it's probably not surprising or too expensive). And then the caller can choose to rewrite the commit structure with the new trees, if so desired.

For similar organizational purposes, I think the callee should be responsible for procuring the working copy to operate on.

Reply:
Ok, I've removed the `rewritten_id` field, so the caller is responsible for creating it.